shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -1080,22 +1078,62 @@ routing message to a random node so that:
-                        &self.sasl_mechanism,
-                        &self.nodes,
-                        destination,
-                        err.clone().into(),
-                    )
-                    .await?;
-                // If we succesfully recreate the outgoing connection we still need to terminate this incoming connection since the request is lost.
-                return Err(err.into());
+                Ok(connection) => {
+                    if let Err(err) = connection.send(requests.requests) {
+                        // Dont retry the send on the new connection since we cant tell if the broker received the request or not.
+                        self.connections
+                            .handle_connection_error(
+                                &self.connection_factory,
+                                &self.authorize_scram_over_mtls,
+                                &self.sasl_mechanism,
+                                &self.nodes,
+                                destination,
+                                err.clone().into(),
+                            )
+                            .await?;
+                        // If we succesfully recreate the outgoing connection we still need to terminate this incoming connection since the request is lost.
+                        return Err(err.into());
+                    }
+                }
+                Err(err) => {
+                    // set node as down, the connection already failed to create so no point running through handle_connection_error,
+                    // as that will recreate the connection which we already know just failed.
+                    // Instead just directly set the node as down and return the error
+
+                    // set node as down
+                    self.nodes
+                        .iter()
+                        .find(|x| match destination {
+                            Destination::Id(id) => x.broker_id == id,
+                            Destination::ControlConnection => {
+                                &x.kafka_address
+                                    == self
+                                        .connections
+                                        .control_connection_address
+                                        .as_ref()
+                                        .unwrap()
+                            }
+                        })
+                        .unwrap()
+                        .set_state(NodeState::Down);
+
+                    // bubble up error
+                    let request_types: Vec<String> = requests
+                        .requests
+                        .iter_mut()
+                        .map(|x| match x.frame() {
+                            Some(Frame::Kafka(KafkaFrame::Request { header, .. })) => {
+                                format!("{:?}", ApiKey::try_from(header.request_api_key).unwrap())
+                            }
```

Comment:
Why do we log the api key here?

Reply:
API key is the message type. E.g. fetch, produce, metadata
I included it since I found it useful for debugging routing issues that were causing connections to be opened against down nodes.

Reply:
Right, just read https://kafka.apache.org/0101/protocol.html#protocol_api_keys and found out Kafka defines the request type using "API key" which is a bit confusing because I thought it's a credential.

Reply:
oh yeah, that would be very concerning :laughing: